src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/oops/Field.java

```diff
@@ -40,23 +41,79 @@ public class Field {
 
   /** Constructor for fields that are named in an InstanceKlass's
       fields array (i.e., named, non-VM fields) */
-  Field(InstanceKlass holder, int fieldIndex) {
+  private Field(InstanceKlass holder, int fieldIndex, FieldInfoValues values) {
     this.holder = holder;
     this.fieldIndex = fieldIndex;
+    this.values = values;
+    offset = values.offset;
 
-    offset               = holder.getFieldOffset(fieldIndex);
-    genericSignature     = holder.getFieldGenericSignature(fieldIndex);
-
-    name                 = holder.getFieldName(fieldIndex);
+    name = holder.getSymbolFromIndex(values.nameIndex, isInjected());
+    signature = holder.getSymbolFromIndex(values.signatureIndex, isInjected());
     id          = new NamedFieldIdentifier(name.asString());
-
-    signature            = holder.getFieldSignature(fieldIndex);
     fieldType   = new FieldType(signature);
+    accessFlags = new AccessFlags(values.accessFlags);
+
+    if (isGeneric()) {
+      genericSignature = holder.getSymbolFromIndex(values.genericSignatureIndex, isInjected());
+    }
+  }
 
-    short access         = holder.getFieldAccessFlags(fieldIndex);
-    accessFlags = new AccessFlags(access);
+  /** Constructor for cloning an existing Field object */
+  Field(InstanceKlass holder, int fieldIndex) {
+      this(holder, fieldIndex, holder.getField(fieldIndex).values);
   }
 
+
+  static class FieldInfoValues {
+    int nameIndex;
+    int signatureIndex;
+    int offset;
+    int accessFlags;
+    int fieldFlags;
+    int initialValueIndex;
+    int genericSignatureIndex;
+    int contendedGroup;
```

Comment:
It seems that these should all be shorts. All the getter methods are casting them to short.